lib/spack/spack/package_base.py

```diff
@@ -25,7 +25,23 @@
 import traceback
 import typing
 import warnings
-from typing import Any, Callable, Dict, Iterable, List, Optional, Set, Tuple, Type, TypeVar, Union
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+if TYPE_CHECKING:
+    from typeshed import SupportsRichComparisonT
```

Comment:
Where do we get the `typeshed` import from?

Reply:
It’s part of `mypy` and has to be there if you’re running `mypy`, so I think it’s escape to use it conditionally like this.

See https://github.com/python/mypy/tree/master/mypy

Reply:
Can we use something based on `typing.Protocol` instead? That would make us independent of `mypy` in the annotations, and we could avoid conditional imports.

Reply:
Also, I think using the above, or even my suggestion, makes style checks possible only for Python 3.8 and above.

Reply:
`typing.Protocol` is also only in 3.8, which is why I suspect `SupportsRichComparisonT` doesn't exist until then -- it's a protocol.  I'm not sure how we should do this other than vendor `typing_extensions`, which is maybe a good idea anyway.  It would get us typed dicts, literals and other stuff I've wanted in the past.

Ok, I vendored `typing_extensions` and can use `Protocol` -- that was easy and it's only one file. Of course the annoying part is that even in newer pythons, protocols like `SupportsRichComparisonT` isn't defined outside `typeshed`, and googling this tells me that I'm not supposed to import directly from `typeshed`... so I have to make my own version of the protocol even though it's the thing that `mypy` uses to describe types (I found it because this was the type bound it was complaining about).

What's the typical way to deal with this?  Do people just make their own protocols?  Maybe @adamjstewart knows.

Reply:
Ok I implemented this with a protocol -- see what you think.